lib/logstash/inputs/sqs.rb

```diff
@@ -76,101 +77,108 @@ class LogStash::Inputs::SQS < LogStash::Inputs::Threadable
-      @sqs_queue = @sqs.queues.named(@queue)
-      @logger.info("Connected to AWS SQS queue successfully.", :queue => @queue)
-    rescue Exception => e
-      @logger.error("Unable to access SQS queue.", :error => e.to_s, :queue => @queue)
-      throw e
-    end # begin/rescue
-  end # def register
+  def aws_sqs_client
+    @sqs ||= Aws::SQS::Client.new(aws_options_hash)
+  end
 
-  public
-  def run(output_queue)
-    @logger.debug("Polling SQS queue", :queue => @queue)
+  def queue_url
+    @queue_url ||= begin
+                     aws_sqs_client.get_queue_url(:queue_name =>  @queue)[:queue_url]
+                   rescue Aws::SQS::Errors::ServiceError => e
+                     @logger.error("Cannot establish connection to Amazon SQS", :error => e)
+                     raise LogStash::ConfigurationError, "Verify the SQS queue name and your credentials"
+                   end
+  end
+
+  def verify_credentials!
+    queue_url
+  end
+
+  def poller
+    Aws::SQS::QueuePoller.new(queue_url, :client => aws_sqs_client)
+  end
 
-    receive_opts = {
-        :limit => 10,
-        :visibility_timeout => 30,
-        :attributes => [:sent_at]
+  def polling_options
+    { 
+      :max_number_of_messages => MAX_MESSAGES_TO_FETCH,
+      :attribute_names => ["SentTimestamp"],
     }
+  end
 
-    continue_polling = true
-    while running? && continue_polling
-      continue_polling = run_with_backoff(60, 1) do
-        @sqs_queue.receive_message(receive_opts) do |message|
-          if message
-            @codec.decode(message.body) do |event|
-              decorate(event)
-              if @id_field
-                event[@id_field] = message.id
-              end
-              if @md5_field
-                event[@md5_field] = message.md5
-              end
-              if @sent_timestamp_field
-                event[@sent_timestamp_field] = LogStash::Timestamp.new(message.sent_timestamp).utc
-              end
-              @logger.debug? && @logger.debug("Processed SQS message", :message_id => message.id, :message_md5 => message.md5, :sent_timestamp => message.sent_timestamp, :queue => @queue)
-              output_queue << event
-              message.delete
-            end # codec.decode
-          end # valid SQS message
-        end # receive_message
-      end # run_with_backoff
-    end # polling loop
-  end # def run
-
-  def teardown
-    @sqs_queue = nil
-    finished
-  end # def teardown
+  def decode_event(message)
+    @codec.decode(message.body) do |event|
+      return event
+    end
+  end
+  
+  def add_sqs_data(event, message)
+    event[@id_field] = message.message_id if @id_field
+    event[@md5_field] = message.md5_of_body if @md5_field
+    event[@sent_timestamp_field] = LogStash::Timestamp.new(message.attributes.sent_timestamp) if @sent_timestamp_field
+
+    return event
+  end
+
+  def handle_message(message)
+    event = decode_event(message)
+    decorate(event)
```

Comment:
I think I'd expect that if I use

```
sqs {
  ..
  ..
  add_field => ["test", "test"]
}
```
Field `test` will always have `test` in it when leaving the plugin, but if you specify `id_field` to be "test" then it will be replaced since `decorate` happens before `add_sqs_data`.

Thoughts?